### PR TITLE
New changes

### PR DESCRIPTION
**Pull request by IlchukPasha**

### CODE REVIEW

File: app/controllers/auth.js [hunk 1]
```diff
@@ -44,26 +43,23 @@ router.post('/signup', function(req, res, next) {
...
-        return next(err);
-      }
-      let user = {
-        first_name: req.body.first_name,
-        last_name: req.body.last_name,
-        email: req.body.email,
-        password: bcrypt.hashSync(req.body.password, salt),
-        role_id: role.id
-      };
+    console.log('passses');
```
Comment (mazahell) [inline app/controllers/auth.js]: Remove all console logs!

File: app/controllers/user.js [hunk 2]
```diff
@@ -3,20 +3,25 @@ const router = express.Router();
...
     if (err) {
       return next(err);
     }
     res.json(users);
   });
 });
 
-router.get('/:id', checkUserAdminMdwr('user'), function (req, res, next) {
-  User.getById(req.params.id, function (err, user) {
+router.get('/:id', roleMdwr(['admin', 'user']), current_user_mdwr, (req, res, next) => {
```
Comment (mazahell) [inline app/controllers/user.js]: why two roles?

File: app/middlewares/current_user.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
+const { findIndex } = require('lodash');
```
Comment (mazahell) [inline app/middlewares/current_user.js]: bad logic

File: app/middlewares/image_upload.js [hunk 4]
```diff
@@ -3,7 +3,8 @@ const fs = require('fs');
 const Tweet = require('./../models/tweet');
```
Comment (mazahell) [inline app/middlewares/image_upload.js]: bad logic

File: app/middlewares/tweet_owner.js [hunk 5]
```diff
@@ -1,4 +1,5 @@
 const knex = require('./../libs/knex');
```
Comment (mazahell) [inline app/middlewares/tweet_owner.js]: bad logic

File: app/middlewares/validators/Validator.js [hunk 6]
```diff
@@ -13,6 +15,22 @@ Validator.registerAsync('unique', function(email, attribute, req, passes) {
     });
 });
 
+Validator.registerAsync('unique_except_current_user', function(email, attribute, req, passes) {
+  console.log(req.params);
```
Comment (mazahell) [inline app/middlewares/validators/Validator.js]: remove all console logs

File: app/middlewares/validators/Validator.js [hunk 7]
```diff
@@ -39,17 +57,17 @@ Validator.registerAsync('tweet_exist', function(tweet_id, attribute, req, passes
...
-    .where('r.id', role_id)
-    .first()
-    .count('* as c')
-    .then(function(role) {
-      passes(role.c === 1, 'This role not exist.');
-    })
-    .catch(function(err) {
-      passes(false, 'error in validation');
-    });
+Validator.registerAsync('roles_exist', function(roles, attribute, req, passes) {
```
Comment (mazahell) [inline app/middlewares/validators/Validator.js]: very bad logic

File: app/middlewares/validators/user_update.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+module.exports = (req, res, next) => {
+  let rules = {
+    email: 'required|email|unique_except_current_user',
+    password: 'required|min:5',
+    first_name: 'required|min:5|max:30',
+    last_name: 'required|min:5|max:30'
+  };
+
+  if (req.body.roles) {
```
Comment (mazahell) [inline app/middlewares/validators/user_update.js]: and for admin user?

File: app/models/user.js [hunk 9]
```diff
@@ -83,22 +84,49 @@ var User = bookshelf.Model.extend(
...
           'u.id as id',
           'u.email as email',
-          'r.title as role',
+          'u.roles as roles',
           'u.first_name as first_name',
           'u.last_name as last_name'
         )
-        .innerJoin('roles as r', 'u.role_id', 'r.id')
         .then(users => {
+          if (users) {
```
Comment (mazahell) [inline app/models/user.js]: what?

File: db/migrations/20170712122159_change_users_table.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
+exports.up = function(knex, Promise) {
+  return knex.schema.table('users', table => {
+    table.dropForeign('role_id');
+    table.dropColumn('role_id');
+    table.string('roles', 20).notNullable();
```
Comment (mazahell) [inline db/migrations/20170712122159_change_users_table.js]: 20 is very small

File: app/middlewares/validators/Validator.js (hunk 7 above)
Comment (IlchukPasha) [inline app/middlewares/validators/Validator.js]: what exactly?

Comment (IlchukPasha) [inline app/middlewares/validators/Validator.js]: no need to check if roles are string?

File: app/middlewares/current_user.js (hunk 3 above)
Comment (IlchukPasha) [inline app/middlewares/current_user.js]: why?

Comment (mazahell) [inline app/middlewares/current_user.js]: some parts of code usage twice

Comment (IlchukPasha) [inline app/middlewares/current_user.js]: which parts? can you explain?

**Pull request merged**